### PR TITLE
feat(general): add support for Python 3.11

### LINKED ISSUE #2732: Maintenance: Support for Python 3.11
### Why is this needed?

Python 3.11 official docker image is available in **preview**: https://github.com/aws/aws-lambda-base-images/issues/62#issuecomment-1569890315

### Which area does this relate to?

Other

### Solution

Add support in CI. We have some customers using Python 3.11 using OCI images in Lambda.

### Acknowledgment

- [X] This request meets [Powertools for AWS Lambda (Python) Tenets](https://docs.powertools.aws.dev/lambda/python/latest/#tenets)
- [ ] Should this be considered in other Powertools for AWS Lambda languages? i.e. [Java](https://github.com/aws-powertools/powertools-lambda-java/), [TypeScript](https://github.com/aws-powertools/powertools-lambda-typescript/), and [.NET](https://github.com/aws-powertools/powertools-lambda-dotnet/)

### PR DESCRIPTION
**Pull request by rubenfonseca**
**Issue number:** #2732

## Summary

This PR adds support for Python 3.11.

Tasks to complete before merging this PR

- [x] Fix problems with snappy dev dependency
- [x] Fix problems with fastjsonschema v2.18.0
- [x] Fix problems with LRU tests
- [x] Validate all features supported by Powertools using AWS Lambda Python3.11 image
- [x] Validate Pydanticv2 using AWS Lambda Python3.11 image
- [ ] Update all docs examples to use python3.11 instead of python3.10
- [x] Run e2e test

## Checklist

If your change doesn't seem to apply, please leave them unchecked.

- [x] [Meet tenets criteria](https://docs.powertools.aws.dev/lambda/python/#tenets)
- [x] I have performed a self-review of this change
- [x] Changes have been tested
- [x] Changes are documented
- [x] PR title follows [conventional commit semantics](https://github.com/aws-powertools/powertools-lambda-python/blob/develop/.github/semantic.yml)

<details>
<summary>Is this a breaking change?</summary>

**RFC issue number**:

Checklist:

- [ ] Migration process documented
- [ ] Implement warnings (if it can live side by side)

</details>

## Acknowledgment

By submitting this pull request, I confirm that you can use, modify, copy, and redistribute this contribution, under the terms of your choice.

**Disclaimer**: We value your time and bandwidth. As such, any pull requests created on non-triaged issues might not be successful.

### CODE REVIEW

**Review by heitorlessa — APPROVED** [review]

**Comment by rubenfonseca** [comment]
TODO: remove snappy dependency and all references in examples and documentation /cc @leandrodamascena

File: tests/functional/validator/test_validator.py
Comment (leandrodamascena) [inline tests/functional/validator/test_validator.py:33]: fastjsonschema changed the way of reporting bugs in the latest version (2.18.0). Now it just reports missing fields instead of all fields.

https://github.com/horejsek/python-fastjsonschema/commit/756540088687cda351390f687b92e602feaa7dc6

> * Improved error message for required props - only missing are reported

File: tests/unit/test_lru_cache.py
Comment (leandrodamascena) [inline tests/unit/test_lru_cache.py:69]: NOTE: I removed some characters to allow for adding a comment.

First time I had contact with this and the test is falling. I understand why it's failing in Python 3.11, but I have no idea why we need it here. Do you have any information about this worth sharing @heitorlessa?

ERROR:

>          > pytest.fail("GitHub #300: LRUDict pop bug has been fixed.")
>          E Failed: GitHub #300: LRUDict pop bug has been fixed.

File: pyproject.toml
Comment (leandrodamascena) [inline pyproject.toml:58]: I'll recreate the example that use this library.

**Review by leandrodamascena — CHANGES_REQUESTED** [review]
I added some important information I found on this PR before we merged.

File: pyproject.toml
Comment (leandrodamascena) [reply pyproject.toml:58]: The example is using zlib instead snappy.

**Comment by leandrodamascena** [comment]
Can't update the examples because cfn-lint still doesn't offer support to Python 3.11

```sh
linting::cloudformation..................................................Failed
- hook id: cloudformation
- exit code: 2

E3030 You must specify a valid value for Runtime (python3.11). Valid values are ["dotnet6", "dotnetcore1.0", "dotnetcore2.0", "dotnetcore2.1", "dotnetcore3.1", "go1.x", "java11", "java17", "java8", "java8.al2", "nodejs", "nodejs10.x", "nodejs12.x", "nodejs14.x", "nodejs16.x", "nodejs18.x", "nodejs4.3", "nodejs4.3-edge", "nodejs6.10", "nodejs8.10", "provided", "provided.al2", "python2.7", "python3.10", "python3.6", "python3.7", "python3.8", "python3.9", "ruby2.5", "ruby2.7", "ruby3.2"]
examples/batch_processing/sam/dynamodb_batch_processing.yaml:17:3
```

There is an issue to add support for this - https://github.com/aws-cloudformation/cfn-lint/issues/2788

**Comment by leandrodamascena** [comment]
Hello @rubenfonseca! This is a HUGE amount of incredible WORK! We can merge this PR merge after @heitorlessa help us with a small problem in one of our tests.

Congrats dude, this is an essential step for us. :rocket: :heart:

File: tests/unit/test_lru_cache.py
Comment (heitorlessa) [reply tests/unit/test_lru_cache.py:69]: That was an old bug in the implementation of LRUDict from `OrderedDict`. Let's skip for 3.11+ until we have more quality time to dive into refactoring this.

Comment (leandrodamascena) [reply tests/unit/test_lru_cache.py:69]: Added a skip for Python 3.11+.

**Review by heitorlessa — APPROVED** [review]

**Comment by heitorlessa** [comment]
Thank you @rubenfonseca and @leandrodamascena - fantastic work ;)

**Review by leandrodamascena — APPROVED** [review]

**Pull request merged**
